File: frontend/scripts/monitor_canary.js
#!/usr/bin/env node

/**
 * Canary Deployment Monitoring Script for MassUGC Studio
 * 
 * Monitors the health of canary deployments by checking:
 * - Download endpoint availability
 * - Error rates from telemetry
 * - User feedback signals
 * - Performance metrics
 * 
 * Usage:
 *   node scripts/monitor_canary.js --platform windows --duration 300 --threshold 5
 * 
 * Options:
 *   --platform      Platform to monitor (windows, macos, all) [default: all]
 *   --duration      Monitoring duration in seconds [default: 300]
 *   --threshold     Error rate threshold percentage [default: 5.0]
 *   --interval      Check interval in seconds [default: 60]
 *   --url           Base URL for canary deployment [default: https://canary.massugc.studio]
 *   --telemetry     Telemetry API endpoint [optional]
 *   --slack-webhook Slack webhook for notifications [optional]
 */

const https = require('https');
const http = require('http');
const fs = require('fs');
const path = require('path');

class CanaryMonitor {
  constructor(options = {}) {
    this.platform = options.platform || 'all';
    this.duration = parseInt(options.duration) || 300; // 5 minutes default
    this.threshold = parseFloat(options.threshold) || 5.0; // 5% error threshold
    this.interval = parseInt(options.interval) || 60; // 1 minute default
    this.baseUrl = options.url || 'https://canary.massugc.studio';
    this.telemetryUrl = options.telemetry;
    this.slackWebhook = options.slackWebhook;
    
    this.startTime = Date.now();
    this.checks = 0;
    this.failures = 0;
    this.metrics = {
      windows: { available: 0, errors: 0, totalChecks: 0 },
      macos: { available: 0, errors: 0, totalChecks: 0 }
    };
  }

  /**
   * Start monitoring the canary deployment
   */
  async start() {
    console.log('🔍 Canary Deployment Monitor Starting');
    console.log(`📊 Configuration:`);
    console.log(`   Platform: ${this.platform}`);
    console.log(`   Duration: ${this.duration}s (${Math.floor(this.duration / 60)} minutes)`);
    console.log(`   Interval: ${this.interval}s`);
    console.log(`   Threshold: ${this.threshold}%`);
    console.log(`   Base URL: ${this.baseUrl}`);
    console.log('');

    const endTime = this.startTime + (this.duration * 1000);
    let checkCount = 0;

    while (Date.now() < endTime) {
      checkCount++;
      const elapsed = Math.floor((Date.now() - this.startTime) / 1000);
      const remaining = Math.floor((endTime - Date.now()) / 1000);
      
      console.log(`\n[Check ${checkCount}/${Math.ceil(this.duration / this.interval)}] Elapsed: ${elapsed}s | Remaining: ${remaining}s`);
      console.log('─'.repeat(70));

      try {
        await this.performHealthChecks();
        await this.checkTelemetry();
        this.displayStatus();
        
        // Check if we've exceeded error threshold
        if (this.getErrorRate() > this.threshold) {
          await this.handleThresholdExceeded();
          return false;
        }
      } catch (error) {
        console.error(`❌ Check failed:`, error.message);
        this.failures++;
      }

      // Wait for next interval (unless this is the last check)
      if (Date.now() < endTime) {
        const waitTime = Math.min(this.interval * 1000, endTime - Date.now());
        await this.sleep(waitTime);
      }
    }

    console.log('\n' + '='.repeat(70));
    console.log('🎉 Monitoring Complete');
    console.log('='.repeat(70));
    this.displayFinalReport();
    
    return this.getErrorRate() <= this.threshold;
  }

  /**
   * Perform health checks on download endpoints
   */
  async performHealthChecks() {
    const platforms = this.platform === 'all' ? ['windows', 'macos'] : [this.platform];

    for (const platform of platforms) {
      this.metrics[platform].totalChecks++;
      
      const downloadUrl = `${this.baseUrl}/downloads/${platform}`;
      console.log(`🔎 Checking ${platform} endpoint: ${downloadUrl}`);

      try {
        const available = await this.checkEndpoint(downloadUrl);
        
        if (available) {
          this.metrics[platform].available++;
          console.log(`   ✅ ${platform}: AVAILABLE`);
        } else {
          this.metrics[platform].errors++;
          console.log(`   ❌ ${platform}: UNAVAILABLE`);
        }
      } catch (error) {
        this.metrics[platform].errors++;
        console.log(`   ❌ ${platform}: ERROR - ${error.message}`);
      }
    }
  }

  /**
   * Check if an endpoint is available
   */
  checkEndpoint(url) {
    return new Promise((resolve) => {
      const protocol = url.startsWith('https') ? https : http;
      const urlObj = new URL(url);

      const options = {
        hostname: urlObj.hostname,
        port: urlObj.port || (protocol === https ? 443 : 80),
        path: urlObj.pathname,
        method: 'HEAD',
        timeout: 10000
      };

      const req = protocol.request(options, (res) => {
        resolve(res.statusCode >= 200 && res.statusCode < 400);
      });

      req.on('error', () => resolve(false));
      req.on('timeout', () => {
        req.destroy();
        resolve(false);
      });

      req.end();
    });
  }

  /**
   * Check telemetry for error rates
   */
  async checkTelemetry() {
    if (!this.telemetryUrl) {
      return;
    }

    console.log(`📊 Checking telemetry: ${this.telemetryUrl}`);

    try {
      const data = await this.fetchJson(this.telemetryUrl);
      
      if (data && data.errorRate !== undefined) {
        const errorRate = parseFloat(data.errorRate);
        console.log(`   📈 Error Rate: ${errorRate.toFixed(2)}%`);
        
        if (errorRate > this.threshold) {
          console.log(`   ⚠️  Warning: Error rate exceeds threshold!`);
        }
      }
    } catch (error) {
      console.log(`   ⚠️  Could not fetch telemetry: ${error.message}`);
    }
  }

  /**
   * Fetch JSON from URL
   */
  fetchJson(url) {
    return new Promise((resolve, reject) => {
      const protocol = url.startsWith('https') ? https : http;
      
      protocol.get(url, (res) => {
        let data = '';
        res.on('data', chunk => data += chunk);
        res.on('end', () => {
          try {
            resolve(JSON.parse(data));
          } catch {
            reject(new Error('Invalid JSON response'));
          }
        });
      }).on('error', reject);
    });
  }

  /**
   * Display current status
   */
  displayStatus() {
    console.log('\n📊 Current Status:');
    
    const platforms = this.platform === 'all' ? ['windows', 'macos'] : [this.platform];
    
    for (const platform of platforms) {
      const m = this.metrics[platform];
      const availability = m.totalChecks > 0 
        ? ((m.available / m.totalChecks) * 100).toFixed(1)
        : 0;
      
      console.log(`   ${platform.padEnd(8)} - Availability: ${availability}% (${m.available}/${m.totalChecks})`);
    }

    const overallErrorRate = this.getErrorRate();
    const status = overallErrorRate <= this.threshold ? '✅ PASS' : '❌ FAIL';
    console.log(`\n   Overall Error Rate: ${overallErrorRate.toFixed(2)}% - ${status}`);
  }

  /**
   * Calculate overall error rate
   */
  getErrorRate() {
    const totalChecks = this.metrics.windows.totalChecks + this.metrics.macos.totalChecks;
    const totalErrors = this.metrics.windows.errors + this.metrics.macos.errors;
    
    return totalChecks > 0 ? (totalErrors / totalChecks) * 100 : 0;
  }

  /**
   * Handle threshold exceeded
   */
  async handleThresholdExceeded() {
    console.log('\n' + '='.repeat(70));
    console.log('🚨 ERROR THRESHOLD EXCEEDED');
    console.log('='.repeat(70));
    
    const errorRate = this.getErrorRate();
    console.log(`Current error rate: ${errorRate.toFixed(2)}%`);
    console.log(`Threshold: ${this.threshold}%`);
    console.log('\n⚠️  Canary deployment should be CANCELLED');

    await this.sendNotification('🚨 Canary Deployment Failed', 
      `Error rate ${errorRate.toFixed(2)}% exceeds threshold ${this.threshold}%`);

    // Create failure report
    this.createReport('FAILED');
  }

  /**
   * Display final report
   */
  displayFinalReport() {
    const duration = Math.floor((Date.now() - this.startTime) / 1000);
    const errorRate = this.getErrorRate();
    const status = errorRate <= this.threshold ? 'PASSED' : 'FAILED';

    console.log(`\n📋 Final Report:`);
    console.log(`   Duration: ${duration}s (${Math.floor(duration / 60)}m ${duration % 60}s)`);
    console.log(`   Total Checks: ${this.checks}`);
    console.log(`   Error Rate: ${errorRate.toFixed(2)}%`);
    console.log(`   Status: ${status}`);
    console.log('');
    
    console.log('📊 Platform Details:');
    const platforms = this.platform === 'all' ? ['windows', 'macos'] : [this.platform];
    
    for (const platform of platforms) {
      const m = this.metrics[platform];
      const availability = m.totalChecks > 0 
        ? ((m.available / m.totalChecks) * 100).toFixed(1)
        : 0;
      
      console.log(`   ${platform}:`);
      console.log(`     - Checks: ${m.totalChecks}`);
      console.log(`     - Available: ${m.available} (${availability}%)`);
      console.log(`     - Errors: ${m.errors}`);
    }

    // Create success report
    this.createReport(status);

    if (status === 'PASSED') {
      console.log('\n✅ Canary deployment is healthy and ready for production promotion');
      this.sendNotification('✅ Canary Deployment Passed', 
        `Error rate ${errorRate.toFixed(2)}% is within threshold ${this.threshold}%`);
    }
  }

  /**
   * Create monitoring report file
   */
  createReport(status) {
    const report = {
      timestamp: new Date().toISOString(),
      status,
      duration: Math.floor((Date.now() - this.startTime) / 1000),
      errorRate: this.getErrorRate(),
      threshold: this.threshold,
      platform: this.platform,
      metrics: this.metrics
    };

    const reportPath = path.join(process.cwd(), 'canary-monitor-report.json');
    fs.writeFileSync(reportPath, JSON.stringify(report, null, 2));
    console.log(`\n📄 Report saved to: ${reportPath}`);
  }

  /**
   * Send notification (Slack, email, etc.)
   */
  async sendNotification(title, message) {
    if (!this.slackWebhook) {
      return;
    }

    console.log(`\n📢 Sending notification: ${title}`);

    try {
      const payload = JSON.stringify({
        text: `${title}\n${message}`,
        blocks: [
          {
            type: 'header',
            text: { type: 'plain_text', text: title }
          },
          {
            type: 'section',
            text: { type: 'mrkdwn', text: message }
          }
        ]
      });

      await this.postWebhook(this.slackWebhook, payload);
      console.log('   ✅ Notification sent');
    } catch (error) {
      console.log(`   ⚠️  Failed to send notification: ${error.message}`);
    }
  }

  /**
   * Post to webhook
   */
  postWebhook(url, data) {
    return new Promise((resolve, reject) => {
      const urlObj = new URL(url);
      const options = {
        hostname: urlObj.hostname,
        port: urlObj.port || 443,
        path: urlObj.pathname,
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Content-Length': Buffer.byteLength(data)
        }
      };

      const req = https.request(options, (res) => {
        res.on('data', () => {});
        res.on('end', () => resolve());
      });

      req.on('error', reject);
      req.write(data);
      req.end();
    });
  }

  /**
   * Sleep utility
   */
  sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
  }
}

// Parse command-line arguments
function parseArgs() {
  const args = process.argv.slice(2);
  const options = {};

  for (let i = 0; i < args.length; i += 2) {
    const key = args[i].replace(/^--/, '');
    const value = args[i + 1];
    options[key] = value;
  }

  return options;
}

// Main execution
if (require.main === module) {
  const options = parseArgs();
  const monitor = new CanaryMonitor(options);

  monitor.start()
    .then(success => {
      process.exit(success ? 0 : 1);
    })
    .catch(error => {
      console.error('❌ Monitoring failed:', error);
      process.exit(1);
    });
}

module.exports = CanaryMonitor;

File: backend/scripts/monitor_canary.py
#!/usr/bin/env python3
"""
Canary Deployment Monitor
Monitors canary deployment health and triggers rollback if issues detected
"""

import requests
import time
import sys
import json
from datetime import datetime, timedelta
from typing import Dict, List, Tuple

class CanaryMonitor:
    def __init__(self, base_url: str, duration_seconds: int = 1800):
        self.base_url = base_url.rstrip('/')
        self.duration_seconds = duration_seconds
        self.error_count = 0
        self.total_checks = 0
        self.error_threshold = 5.0  # 5% error rate
        self.checks = []
        
    def check_health(self) -> Tuple[bool, str]:
        """Check health endpoint"""
        try:
            response = requests.get(
                f"{self.base_url}/health",
                timeout=10
            )
            
            if response.status_code == 200:
                return True, "OK"
            else:
                return False, f"HTTP {response.status_code}"
                
        except requests.exceptions.Timeout:
            return False, "Timeout"
        except requests.exceptions.ConnectionError:
            return False, "Connection Error"
        except Exception as e:
            return False, str(e)
    
    def check_metrics(self) -> Dict:
        """Get metrics from endpoint"""
        try:
            response = requests.get(
                f"{self.base_url}/metrics",
                timeout=10
            )
            
            if response.status_code == 200:
                return response.json()
            else:
                return {}
                
        except Exception as e:
            print(f"⚠️  Failed to fetch metrics: {e}")
            return {}
    
    def calculate_error_rate(self) -> float:
        """Calculate current error rate percentage"""
        if self.total_checks == 0:
            return 0.0
        return (self.error_count / self.total_checks) * 100
    
    def should_rollback(self) -> bool:
        """Determine if automatic rollback should be triggered"""
        error_rate = self.calculate_error_rate()
        
        # Trigger rollback if error rate exceeds threshold
        if error_rate > self.error_threshold:
            return True
        
        # Trigger rollback if last 5 consecutive checks failed
        if len(self.checks) >= 5:
            recent_checks = self.checks[-5:]
            if all(not check['success'] for check in recent_checks):
                return True
        
        return False
    
    def log_check(self, success: bool, message: str, metrics: Dict = None):
        """Log check result"""
        check = {
            'timestamp': datetime.now().isoformat(),
            'success': success,
            'message': message,
            'metrics': metrics or {},
            'error_rate': self.calculate_error_rate()
        }
        self.checks.append(check)
        
        # Print status
        status_icon = "✓" if success else "❌"
        print(f"{status_icon} Check {self.total_checks}: {message} (Error rate: {self.calculate_error_rate():.2f}%)")
    
    def save_report(self, filepath: str = "canary_report.json"):
        """Save monitoring report"""
        report = {
            'monitoring_started': self.checks[0]['timestamp'] if self.checks else None,
            'monitoring_ended': self.checks[-1]['timestamp'] if self.checks else None,
            'duration_seconds': self.duration_seconds,
            'total_checks': self.total_checks,
            'error_count': self.error_count,
            'error_rate': self.calculate_error_rate(),
            'error_threshold': self.error_threshold,
            'checks': self.checks
        }
        
        with open(filepath, 'w') as f:
            json.dump(report, f, indent=2)
        
        print(f"📊 Report saved to {filepath}")
    
    def monitor(self) -> int:
        """
        Main monitoring loop
        Returns: 0 for success, 1 for failure
        """
        start_time = datetime.now()
        end_time = start_time + timedelta(seconds=self.duration_seconds)
        
        print("=" * 60)
        print("🔍 CANARY DEPLOYMENT MONITOR")
        print("=" * 60)
        print(f"Base URL: {self.base_url}")
        print(f"Duration: {self.duration_seconds/60:.1f} minutes")
        print(f"Error Threshold: {self.error_threshold}%")
        print(f"Started: {start_time.strftime('%Y-%m-%d %H:%M:%S')}")
        print("=" * 60)
        print()
        
        try:
            while datetime.now() < end_time:
                self.total_checks += 1
                
                # Perform health check
                success, message = self.check_health()
                
                if not success:
                    self.error_count += 1
                
                # Fetch metrics
                metrics = self.check_metrics()
                
                # Log the check
                self.log_check(success, message, metrics)
                
                # Check if rollback should be triggered
                if self.should_rollback():
                    print()
                    print("=" * 60)
                    print("🚨 AUTOMATIC ROLLBACK TRIGGERED")
                    print("=" * 60)
                    print(f"Error rate: {self.calculate_error_rate():.2f}%")
                    print(f"Threshold: {self.error_threshold}%")
                    print(f"Total checks: {self.total_checks}")
                    print(f"Failed checks: {self.error_count}")
                    print()
                    self.save_report()
                    return 1
                
                # Wait before next check
                time.sleep(60)  # Check every minute
        
        except KeyboardInterrupt:
            print("\n\n⚠️  Monitoring interrupted by user")
            self.save_report()
            return 1
        
        # Monitoring complete
        print()
        print("=" * 60)
        print("✅ MONITORING COMPLETE")
        print("=" * 60)
        print(f"Total checks: {self.total_checks}")
        print(f"Failed checks: {self.error_count}")
        print(f"Error rate: {self.calculate_error_rate():.2f}%")
        print(f"Threshold: {self.error_threshold}%")
        print()
        
        final_error_rate = self.calculate_error_rate()
        
        self.save_report()
        
        if final_error_rate > self.error_threshold:
            print("❌ Canary deployment FAILED - Error rate too high")
            return 1
        else:
            print("✅ Canary deployment PASSED - Ready for production")
            return 0


def main():
    import argparse
    
    parser = argparse.ArgumentParser(
        description='Monitor canary deployment health'
    )
    parser.add_argument(
        '--url',
        default='https://deploy.massugc.com/canary',
        help='Base URL for canary deployment'
    )
    parser.add_argument(
        '--duration',
        type=int,
        default=1800,
        help='Monitor duration in seconds (default: 1800 = 30 minutes)'
    )
    parser.add_argument(
        '--error-threshold',
        type=float,
        default=5.0,
        help='Error rate threshold percentage (default: 5.0)'
    )
    parser.add_argument(
        '--report',
        default='canary_report.json',
        help='Output report file path'
    )
    
    args = parser.parse_args()
    
    monitor = CanaryMonitor(args.url, args.duration)
    monitor.error_threshold = args.error_threshold
    
    exit_code = monitor.monitor()
    
    sys.exit(exit_code)


if __name__ == "__main__":
    main()

